pyro/contrib/gp/util.py

```diff
@@ -1,139 +1,246 @@
+        guide = dist_constructor(**dist_args)
+
+        # no need to do transforms when support is real
+        if self._priors[name].support is constraints.real:
+            return pyro.sample(name, guide.independent())
+
+        # otherwise, we do inference in unconstrained space and transform the value
+        # back to original space
+        unconstrained_value = pyro.sample("{}_latent".format(name), guide.independent(),
+                                          infer={"is_auxiliary": True})
+        transform = biject_to(self._priors[name].support)
+        value = transform(unconstrained_value)
+        log_density = transform.inv.log_abs_det_jacobian(value, unconstrained_value)
+        p = pyro.sample(name, dist.Delta(value, log_density.sum(), event_dim=value.dim()))
+        return p
+
+    def _register_param(self, name):
         """
-        Gets the current value of a parameter. The correct behavior will depend on
-        ``mode`` of this object (see :meth:`set_mode` method).
+        In "model" mode, lifts the parameter with name ``name`` to a random sample using
+        a predefined prior (from :meth:`set_prior` method).
 
-        :param str param: Name of the parameter.
+        :param str name: Name of the parameter.
         """
-        if param not in self._registered_params:  # set_mode() has not been called yet
-            return getattr(self, param)
-        else:
-            return self._registered_params[param]
-
-    def _register_param(self, param, mode="model"):
-        """
-        Registers a parameter to Pyro. It can be seen as a wrapper for
-        :func:`pyro.param` and :func:`pyro.sample` primitives.
-
-        :param str param: Name of the parameter.
-        :param str mode: Either "model" or "guide".
-        """
-        if param in self._fixed_params:
-            self._registered_params[param] = self._fixed_params[param]
-            return
-        prior = self._priors.get(param)
-        if self.name is None:
-            param_name = param
-        else:
-            param_name = param_with_module_name(self.name, param)
-
-        if prior is None:
-            constraint = self._constraints.get(param)
-            default_value = getattr(self, param)
-            if constraint is None:
-                p = pyro.param(param_name, default_value)
-            else:
-                p = pyro.param(param_name, default_value, constraint=constraint)
-        elif mode == "model":
-            p = pyro.sample(param_name, prior)
-        else:  # prior != None and mode = "guide"
-            MAP_param_name = param_name + "_MAP"
-            # initiate randomly from prior
-            MAP_param = pyro.param(MAP_param_name, prior)
-            p = pyro.sample(param_name, dist.Delta(MAP_param))
-
-        self._registered_params[param] = p
+        if name in self._priors:
+            with autoname.scope(prefix=self._get_name()):
```

Comment:
The `scope` is put here instead of `pyro.param_with_module_name(module_name, param_name)` for simplicity, and to compatible with `autoname.name_count` decorator.